### ISSUE TITLE
Cannot pass list of data source map attributes as input to a resource

### ISSUE THREAD

**Issue by victoriaalee** [hashicorp/terraform#21352]
### Terraform Version
```
Terraform v0.11.13
```

### Terraform Configuration Files
```hcl
local {
    columns = "${list(
      map("name", "col0", "type", "boolean"),
      map("name", "col1", "type", "integer")
  )}"
}

data "yaml_map_of_strings" "table_test" {
  input = <<EOF
name: example_name
owner: example_owner
columns:
  - name: col0
    type: boolean
  - name: col1
    type: integer
EOF
}

data "yaml_list_of_strings" "columns" {
  input = "${data.yaml_map_of_strings.table_test.output["columns"]}"
}

data "yaml_map_of_strings" "columns" {
  count = "${length(data.yaml_list_of_strings.columns.output)}"
  input = "${element(data.yaml_list_of_strings.columns.output, count.index)}"
}

resource "aws_glue_catalog_table" "table" {
  name = "table"
  database_name = "dummy_db"
  catalog_id = "${local.catalog_id}"
  storage_descriptor {
    columns = "${local.columns}"
    location = "s3://vl1619-tst/"
  }
}

resource "aws_glue_catalog_table" "table_test" {
  name = "table_test"
  database_name = "dummy_db"
  catalog_id = "${local.catalog_id}"
  storage_descriptor {
    columns = ["${data.yaml_map_of_strings.columns.*.output}"]
    location = "s3://vl1619-tst/"
  }
}

output "columns_works" {
  value = "${local.columns}"
}

output "columns" {
  value = ["${data.yaml_map_of_strings.columns.*.output}"]
}
```

### Expected Behavior
Both resources should have been successfully created.

### Actual Behavior
```
Error: aws_glue_catalog_table.table_test: "storage_descriptor.0.columns.0.name": required field is not set
```

### Steps to Reproduce
`terraform apply`
Note: uses [this](https://github.com/ashald/terraform-provider-yaml) 3rd party plugin

### Additional Context
Relevant definition of the `columns` input field from the AWS provider: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_glue_catalog_table.go#L88-L107

These are the outputs when the `aws_glue_catalog_table.table_test` resource is commented out:
```
Outputs:

columns = [
    {
        name = col0,
        type = boolean
    },
    {
        name = col1,
        type = integer
    }
]
columns_works = [
    {
        name = col0,
        type = boolean
    },
    {
        name = col1,
        type = integer
    }
]
```

### References

#17292 - similar issue but the input field was a set of configuration blocks

**Comment by victoriaalee** [comment]
Ah, I see now what's going on. I misunderstood the definition of the input as a list of maps, but it is a list of configuration blocks. Basically the same problem as the issue linked above ^

**Comment by victoriaalee** [comment]
Hm, actually this still doesn't explain why the manual local list of maps worked... reopening.

**Comment by apparentlymart** [comment]
Hi @victoriaalee! Sorry for this confusing behavior.

I'm pretty confused by this behavior myself, to be honest. My next thing to try would be to remove the extra brackets around `"${data.yaml_map_of_strings.columns.*.output}"`, because they shouldn't be necessary here: the expression inside is already a list anyway. Terraform usually removes those redundant brackets as part of being backward-compatible with Terraform 0.6, but it should not be necessary in Terraform 0.11 and the backward compatibility shim may not be working as expected in this case.

If that doesn't help, I suspect that the language improvements in the forthcoming Terraform v0.12.0 release will fix this bug, since the new language implementation is far more consistent in its handling of lists, maps, and objects. The final release isn't out yet, but it's due soon and [release candidate 1](https://www.hashicorp.com/blog/releasing-terraform-0-12-0-rc1) could allow you to verify the fix. However, it doesn't look like the third-party YAML provider you are using has been updated for v0.12 compatibility, so using the new version may not be possible yet.

Depending on why you were using YAML here, you might find that there's a different way to address your problem in v0.12, since it has a few new capabilities. If you were using YAML only to work around the fact that Terraform 0.11 made it hard to traverse through nested data structures, then your same result could potentially be achieved without YAML at all in v0.12:

```hcl
local {
  table = {
    name  = "example_name"
    owner = "example_owner"
    columns = [
      { name = "col0", type = "boolean" },
      { name = "col1", type = "integer" },
    ]
  }
}

resource "aws_glue_catalog_table" "table" {
  name = "table"
  database_name = "dummy_db"
  catalog_id = local.catalog_id
  storage_descriptor {
    columns = local.table.columns
    location = "s3://vl1619-tst/"
  }
}

resource "aws_glue_catalog_table" "table_test" {
  name = "table_test"
  database_name = "dummy_db"
  catalog_id = local.catalog_id
  storage_descriptor {
    columns = local.table.columns
    location = "s3://vl1619-tst/"
  }
}

output "columns" {
  value = local.table.columns
}
```

If having it in a separate file is the goal, and you're willing to tolerate it being JSON instead of YAML, then Terraform 0.12 also includes a built-in `jsondecode` function that can decode JSON natively. `yamldecode` is likely to follow in a later release, but isn't there yet and not immediately planned.

---

One problem I noticed peeking into the source code for the `aws_glue_catalog_table` resource type is that `columns` is defined as being a nested block rather than an attribute, so as currently specified it would need to use block syntax:

```hcl
columns {
  name = "foo"
  type = "bar"
}
```

Given how `columns` is shown in the provider docs example and given that it has a plural name, I assume the author of this code intended for it to be an attribute instead, and so it may well be redefined as one before final v0.12.0 release, but if it remains as currently defined then using it will require [the `dynamic` block syntax](https://www.terraform.io/docs/configuration/expressions.html#dynamic-blocks) in 0.12 instead:

```hcl
  storage_descriptor {
    dynamic "columns" {
      for_each = local.table.columns
      content {
        name = columns.value.name
        type = columns.value.type
      }
    }
    location = "s3://vl1619-tst/"
  }
```

I left a quick note for the AWS provider team about this to see what they think. I'm not sure then which of these forms will be correct once Terraform v0.12.0 final is released, but this `dynamic "columns"` form is what would be needed at the time I write this, if nothing changes.

**Comment by victoriaalee** [comment]
Removing the extra brackets leads to an error about the value not being a list, if I recall correctly from previous attempts.

I made the decision to have inputs provided in YAML to avoid training non-technical folks on the Terraform syntax; a `yamldecode` function sounds like it would be very helpful in the future. For now, I have taken a different approach of a Python script to parse the YAML file and generate a `.tf.json` file.

I'm not sure if I should close this issue (since it seems that the unexpected behavior isn't already specifically addressed in a future release), so I will leave that at your discretion @apparentlymart.

Thanks for the help!

**Comment by beaudryj** [comment]
@victoriaalee do you have a sample of how your parsing that .tf.json, ⬆️ I am trying to solve this same problem :)

**Comment by apparentlymart** [comment]
Hi again!

At the time I write this, the latest version of Terraform now has [`yamldecode`](https://www.terraform.io/docs/configuration/functions/yamldecode.html) as I suggested previously. While this doesn't directly address whatever strange problem was happening in Terraform 0.11, I expect that was a Terraform 0.11 quirk and that implementing the same use-case using Terraform 0.12 and the `yamldecode` function should work better:

```hcl
local {
  settings = yamldecode(
    <<-EOF
      name: example_name
      owner: example_owner
      columns:
        - name: col0
          type: boolean
        - name: col1
          type: integer
    EOF
  )
}

resource "aws_glue_catalog_table" "table_test" {
  name = "table_test"
  database_name = "dummy_db"
  catalog_id = local.catalog_id
  storage_descriptor {
    dynamic "columns" {
      for_each = local.settings.columns
      content {
        name = columns.name
        type = columns.type
      }
    }
    location = "s3://vl1619-tst/"
  }
}
```

Hopefully the above example is helpful to anyone trying to do something like this with Terraform 0.12.4 or later. If you run into any similar problems with the latest version of Terraform, please open a new issue describing it, since Terraform 0.12 is different enough that the root cause is very unlikely to be the same and so we'd need to debug any other issues independently.

It seems like you found a way to work around the v0.11 limitation, @victoriaalee, and I think the issue should be fixed in Terraform 0.12 by using some newer features as above, so I'm going to close this one out. Thanks for reporting this, and sorry we didn't really get to the bottom of the original issue.
